pkg/server/v3/delta_test.go

```diff
@@ -0,0 +1,413 @@
+	"github.com/envoyproxy/go-control-plane/pkg/cache/v3"
+	rsrc "github.com/envoyproxy/go-control-plane/pkg/resource/v3"
+	"github.com/envoyproxy/go-control-plane/pkg/server/stream/v3"
+	"github.com/envoyproxy/go-control-plane/pkg/server/v3"
+)
+
+func (config *mockConfigWatcher) CreateDeltaWatch(req *discovery.DeltaDiscoveryRequest, st *stream.StreamState) (chan cache.DeltaResponse, func()) {
+	config.deltaCounts[req.TypeUrl] = config.deltaCounts[req.TypeUrl] + 1
+
+	// Create our out watch channel to return with a buffer of one
+	out := make(chan cache.DeltaResponse, 1)
+
+	if len(config.deltaResponses[req.TypeUrl]) > 0 {
+		res := config.deltaResponses[req.TypeUrl][0]
+		// In subscribed, we only want to send back what's changed if we detect changes
+		var subscribed []types.Resource
+
+		// Here we set our current version map to what the stream state says it is
+		versionMap := st.ResourceVersions
+
+		r, _ := res.GetDeltaDiscoveryResponse()
+		if st.IsWildcard {
+			for _, resource := range r.Resources {
+				name := resource.GetName()
+				marshaledResource, _ := cache.MarshalResource(resource)
+				oldV := versionMap[name]
+				if v := cache.HashResource(marshaledResource); v != oldV {
+					versionMap[name] = v
+				}
+				subscribed = append(subscribed, resource)
+			}
+		} else {
+			if len(req.GetResourceNamesSubscribe()) > 0 {
+				for _, resource := range r.Resources {
+					for _, alias := range req.GetResourceNamesSubscribe() {
+						if name := resource.GetName(); name == alias {
+							marshaledResource, _ := cache.MarshalResource(resource)
+							oldV := versionMap[name]
+							if v := cache.HashResource(marshaledResource); v != oldV {
+								versionMap[name] = v
+							}
+							subscribed = append(subscribed, resource)
+						}
+					}
+				}
+			}
+		}
+
+		out <- &cache.RawDeltaResponse{
+			DeltaRequest:      req,
+			Resources:         subscribed,
+			SystemVersionInfo: "",
+			NextVersionMap:    versionMap,
+		}
+	} else if config.closeWatch {
+		close(out)
+	} else {
+		config.deltaWatches += 1
+		return out, func() {
+			close(out)
+			config.deltaWatches -= 1
+		}
+	}
+
+	return out, nil
+}
+
+type mockDeltaStream struct {
+	t         *testing.T
+	ctx       context.Context
+	recv      chan *discovery.DeltaDiscoveryRequest
+	sent      chan *discovery.DeltaDiscoveryResponse
+	nonce     int
+	sendError bool
+	grpc.ServerStream
+}
+
+func (stream *mockDeltaStream) Context() context.Context {
+	return stream.ctx
+}
+
+func (stream *mockDeltaStream) Send(resp *discovery.DeltaDiscoveryResponse) error {
+	// check that nonce is monotonically incrementing
+	stream.nonce = stream.nonce + 1
+	if resp.Nonce != fmt.Sprintf("%d", stream.nonce) {
+		stream.t.Errorf("Nonce => got %q, want %d", resp.Nonce, stream.nonce)
+	}
+	// check resources are non-empty
+	if len(resp.Resources) == 0 {
+		stream.t.Error("Resources => got none, want non-empty")
+	}
+	// check that type URL matches in resources
```

Comment:
`// Check that the per resource TypeURL is correctly set.`

Also maybe move this comment to above the relevant loop